experimental/rocm/array_command_buffer.c

```diff
@@ -0,0 +1,305 @@
+  // TODO: Implement barrier with array edges. Right now all the nodes are
+  // serialized.
+  return iree_ok_status();
+}
+
+static iree_status_t iree_hal_rocm_array_command_buffer_wait_events(
+  iree_hal_command_buffer_t *base_command_buffer, iree_host_size_t event_count, const iree_hal_event_t **events,
+  iree_hal_execution_stage_t source_stage_mask, iree_hal_execution_stage_t target_stage_mask,
+  iree_host_size_t memory_barrier_count, const iree_hal_memory_barrier_t *memory_barriers,
+  iree_host_size_t buffer_barrier_count, const iree_hal_buffer_barrier_t *buffer_barriers) {
+  // TODO: Implement barrier with array edges. Right now all the nodes are
+  // serialized.
+  return iree_ok_status();
+}
+
+static iree_status_t iree_hal_rocm_array_command_buffer_discard_buffer(iree_hal_command_buffer_t *base_command_buffer,
+                                                                       iree_hal_buffer_t *buffer) {
+  // nothing to do.
+  return iree_ok_status();
+}
+
+// Splats a pattern value of 1, 2, or 4 bytes out to a 4 byte value.
+static uint32_t iree_hal_rocm_splat_pattern(const void *pattern, size_t pattern_length) {
+  switch (pattern_length) {
+    case 1: {
+      uint32_t pattern_value = *(const uint8_t *)(pattern);
+      return (pattern_value << 24) | (pattern_value << 16) | (pattern_value << 8) | pattern_value;
+    }
+    case 2: {
+      uint32_t pattern_value = *(const uint16_t *)(pattern);
+      return (pattern_value << 16) | pattern_value;
+    }
+    case 4: {
+      uint32_t pattern_value = *(const uint32_t *)(pattern);
+      return pattern_value;
+    }
+    default:
+      return 0;  // Already verified that this should not be possible.
+  }
+}
+
+static iree_status_t iree_hal_rocm_array_command_buffer_fill_buffer(iree_hal_command_buffer_t *base_command_buffer,
+                                                                    iree_hal_buffer_t *target_buffer,
+                                                                    iree_device_size_t target_offset,
+                                                                    iree_device_size_t length, const void *pattern,
+                                                                    iree_host_size_t pattern_length) {
+  iree_hal_rocm_array_command_buffer_t *command_buffer = iree_hal_rocm_array_command_buffer_cast(base_command_buffer);
+
+  hipDeviceptr_t target_device_buffer =
+    iree_hal_rocm_buffer_device_pointer(iree_hal_buffer_allocated_buffer(target_buffer));
+  target_offset += iree_hal_buffer_byte_offset(target_buffer);
+  uint32_t dword_pattern = iree_hal_rocm_splat_pattern(pattern, pattern_length);
+  hipDeviceptr_t dst = target_device_buffer + target_offset;
+  int value = dword_pattern;
+  size_t sizeBytes = length;
+  // TODO(raikonenfnu) : figure out a way to put fill buffer into kernelArray
+  ROCM_RETURN_IF_ERROR(command_buffer->context->syms, hipMemset(dst, value, sizeBytes), "hipMemset");
```

Comment:
Can you use `hipMemsetAsync` to add it to the stream? Overall I assume you are building a stream with all those command?

Reply:
done